File: 1105_FillingBookcaseShelves.py
from typing import List

class Solution:
    def minHeightShelves(self, books: List[List[int]], shelfWidth: int) -> int:
        n = len(books)
        dp = [0] * (n + 1)

        dp[0] = 0
        dp[1] = books[0][1]

        for i in range(2, n + 1):
            remainingWidth = shelfWidth - books[i - 1][0]
            maxHeight = books[i - 1][1]
            dp[i] = books[i - 1][1] + dp[i - 1]

            j = i - 1
            while j > 0 and remainingWidth - books[j - 1][0] >= 0:
                maxHeight = max(maxHeight, books[j - 1][1])
                remainingWidth -= books[j - 1][0]
                dp[i] = min(dp[i], maxHeight + dp[j - 1])
                j -= 1

        return dp[n]

File: 1105_FillingBookcaseShelves.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {

private:
    unordered_map<string, int> memo;
     
    int dfs(const vector<vector<int>>& books, const int& shelfWidth, int index, int currShelfHeight, int remainingShelfWidth){
        if (index == books.size()){
            return currShelfHeight;
        }

        string key = to_string(index) + "," + to_string(remainingShelfWidth);
        if (memo.find(key) != memo.end()){
            return memo[key];
        }
        
        int placeOnSameShelf = INT_MAX;
        if (remainingShelfWidth >= books[index][0]) 
            placeOnSameShelf = dfs(books, shelfWidth, index+1, max(currShelfHeight,books[index][1]), remainingShelfWidth-books[index][0]); 

        int placeOnNextShelf = INT_MAX;
        if (currShelfHeight != 0)
            placeOnNextShelf = currShelfHeight + dfs(books, shelfWidth, index+1, books[index][1], shelfWidth-books[index][0]);

        return memo[key] = min(placeOnNextShelf, placeOnSameShelf);
    }

public:
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        return dfs(books, shelfWidth, 0, 0, shelfWidth);
    }
};